### PR TITLE
Fix Enumerable.Take/SkipLast on mutable collections

### LINKED ISSUE #39864: IEnumerable<T>.SkipLast and TakeLast behave wrong
### Description

`SkipLast` and `TakeLast` behave wrong when the source is a countable collection such as `List<T>` as it internally determines the min and max index to take which is wrong when the collection is modified afterwards.

### Configuration

Observed in .NET Core 3.1 and .NET 5 preview

### Other information

The defect for `TakeLast` was likely introduced with https://github.com/dotnet/corefx/pull/36051 - everyone was just focused on performance that nobody tested correctness it seems. It was then later also applied to `SkipLast` in https://github.com/dotnet/corefx/pull/41342

Code to repro defect:

```
var numbers = new List<int>(Enumerable.Range(1, 10));
var last5 = numbers.TakeLast(5);
foreach (var i in last5) Console.WriteLine(i);
numbers.AddRange(new[] { 11, 12, 13, 14, 15 });
foreach (var i in last5) Console.WriteLine(i);
Console.ReadLine();
```

### PR DESCRIPTION
**Pull request by stephentoub**
LINQ operators are lazy and can't factor in the current contents of a mutable list when making decisions about how to subsequently process that list.  The optimizations in https://github.com/dotnet/corefx/pull/36051 and https://github.com/dotnet/corefx/pull/41342 violated that and broke TakeLast and SkipLast when the underlying source changes after calling Take/SkipLast and before calling GetEnumerator.

Fixes https://github.com/dotnet/runtime/issues/39864

### CODE REVIEW

File: src/libraries/System.Linq/src/System/Linq/Take.SpeedOpt.cs [hunk 1]
```diff
@@ -12,35 +11,5 @@ public static partial class Enumerable
             source is IPartition<TSource> partition ? partition.Take(count) :
             source is IList<TSource> sourceList ? (IEnumerable<TSource>)new ListPartition<TSource>(sourceList, 0, count - 1) :
             new EnumerablePartition<TSource>(source, 0, count - 1);
-
-        private static IEnumerable<TSource> TakeLastEnumerableFactory<TSource>(IEnumerable<TSource> source, int count)
```
Comment (jkotas) [inline src/libraries/System.Linq/src/System/Linq/Take.SpeedOpt.cs:16]: Delete it from the `Task.SizeOpt.cs` as well?

Comment (stephentoub) [reply src/libraries/System.Linq/src/System/Linq/Take.SpeedOpt.cs:16]: Yup

**Comment by eiriktsarpalis** [comment]
Looks good to me, but wouldn't it in principle be possible to preserve the existing optimization by delaying it to enumerator initialization time? Granted, this might require nontrivial refactoring of the existing implementation.

**Comment by stephentoub** [comment]
> this might require nontrivial refactoring of the existing implementation

Right. That's a separate feature. The faulty optimization is being reverted here, and if someone wants to try again separately and validate it's useful, that's welcome.

**Review by eiriktsarpalis — APPROVED** [review]

**Pull request merged**